### PR TITLE
[#940] feat: Support columnar shuffle with gluten

### PR DESCRIPTION
**Pull request by summaryzb**
### What changes were proposed in this pull request?
support read and write serialized columnar data

### Why are the changes needed?

Fix: #940

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
UnitTest, it's covered by the existing test

### CODE REVIEW

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 1]
```diff
@@ -128,7 +128,11 @@ public WriteBufferManager(
     super(taskMemoryManager, taskMemoryManager.pageSizeBytes(), MemoryMode.ON_HEAP);
     this.bufferSize = bufferManagerOptions.getBufferSize();
     this.spillSize = bufferManagerOptions.getBufferSpillThreshold();
-    this.instance = serializer.newInstance();
+    // columnar shuffle reader use the serialized data directly
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:131]: The serializer is null when we use columnar shuffle, isn't it? Could we add a new test case that the serializer is null.

**Comment by jerqi** [comment]
cc @xianjingfeng

**Comment by xianjingfeng** [comment]
Could you share the changed code in gluten? @summaryzb

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 2]
```diff
@@ -151,25 +154,20 @@ public WriteBufferManager(
...
+   * add serialized columnar data directly when integrate with gluten
+   */
+  public List<ShuffleBlockInfo> addPartitionData(int partitionId, byte[] serializedData) {
+    return addPartitionData(partitionId, serializedData, 0, 0L);
+  }
+
+  public List<ShuffleBlockInfo> addPartitionData(
+          int partitionId, byte[] serializedData, int serializedDataLength, long start) {
+    if (start == 0L) {
+      start = System.currentTimeMillis();
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:167]: Could we assign `start` variable in upper method? It's weird to assign a value here.

Comment (summaryzb) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:167]: agree, so as to `serializedDataLength`

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/reader/RssShuffleReader.java [hunk 3]
```diff
@@ -100,6 +101,7 @@ public RssShuffleReader(
     this.mapStartIndex = mapStartIndex;
     this.mapEndIndex = mapEndIndex;
     this.context = context;
+    this.rssShuffleHandle = rssShuffleHandle;
```
[the remark this answers is not in the document]
Comment (summaryzb) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/reader/RssShuffleReader.java:104]: yes, it's used in `rssShuffleHandle.getNumMaps()`

**Comment by summaryzb** [comment]
> Could you share the changed code in gluten? @summaryzb

Yeah, I'll contribute the patch to gluten today and post the link here later

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/reader/RssShuffleReader.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jerqi) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/reader/RssShuffleReader.java:104]: We have a field `numMap` instead of rssShuffleHanlde.

**Comment by summaryzb** [comment]
> Could you share the changed code in gluten? @summaryzb
PTAL
https://github.com/oap-project/gluten/pull/1966

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 4]
```diff
@@ -141,7 +140,11 @@ public WriteBufferManager(
     this.requireMemoryInterval = bufferManagerOptions.getRequireMemoryInterval();
     this.requireMemoryRetryMax = bufferManagerOptions.getRequireMemoryRetryMax();
     this.arrayOutputStream = new WrappedByteArrayOutputStream(serializerBufferSize);
-    this.serializeStream = instance.serializeStream(arrayOutputStream);
+    // columnar shuffle use the serialized data directly
+    if (serializer != null) {
```
[the remark this answers is not in the document]
Comment (summaryzb) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:144]: > didn't special code in the `addPartitionData` part.

please see here [addPartitionData](https://github.com/oap-project/gluten/blob/374aac4309b66db50d3cadcf4408e9b7f390d9a6/gluten-uniffle/src/main/java/org/apache/spark/shuffle/writer/UniffleColumnarShuffleWriter.java#L273)

**Comment by xianjingfeng** [comment]
> > Could you share the changed code in gluten? @summaryzb
> > PTAL @xianjingfeng @advancedxy @jerqi
> > [oap-project/gluten#1966](https://github.com/oap-project/gluten/pull/1966)

Is it possible that UniffleShuffleManager extends RssShuffleManager and UniffleColumnarShuffleWriter extends RssShuffleWriter?

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (summaryzb) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:144]: follow suggestion

**Review by jerqi — APPROVED** [review]
LGTM, let @xianjingfeng take another look. @xianjingfeng You can @your colleague to ensure that this change is ok for him.

**Comment by summaryzb** [comment]
> Is it possible that UniffleShuffleManager extends RssShuffleManager and UniffleColumnarShuffleWriter extends RssShuffleWriter?

That's really good idea, it'll make the gluten side code more simplify. 
How about refactor that in next patch, since  many fields type need to be changed to protected

**Comment by xianjingfeng** [comment]
> > Is it possible that UniffleShuffleManager extends RssShuffleManager and UniffleColumnarShuffleWriter extends RssShuffleWriter?
> 
> That's really good idea, it'll make the gluten side code more simplify. How about refactor that in next patch, since many fields type need to be changed to protected

It is ok for me.

**Review by xianjingfeng — APPROVED** [review]
LGTM

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 5]
```diff
@@ -128,7 +128,11 @@ public WriteBufferManager(
     super(taskMemoryManager, taskMemoryManager.pageSizeBytes(), MemoryMode.ON_HEAP);
     this.bufferSize = bufferManagerOptions.getBufferSize();
     this.spillSize = bufferManagerOptions.getBufferSpillThreshold();
-    this.instance = serializer.newInstance();
+    // columnar shuffle reader use the serialized data directly
+    if (serializer != null) {
```
[the remark this answers is not in the document]
Comment (summaryzb) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java]: @loukey-lj In writer side the serializer is specify with null directly in gluten side code.  In reader side  it's taken from `rssHandle.getDependency().serializer()`.  Sure serializeStream method is not implemented, because serialize work is done in native engine layer

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 6]
```diff
@@ -151,25 +154,20 @@ public WriteBufferManager(
...
-      serializeStream.writeKey(null, ManifestFactory$.MODULE$.Null());
-    }
-    if (value != null) {
-      serializeStream.writeValue(value, ClassTag$.MODULE$.apply(value.getClass()));
-    } else {
-      serializeStream.writeValue(null, ManifestFactory$.MODULE$.Null());
+  /**
+   * add serialized columnar data directly when integrate with gluten
+   */
+  public List<ShuffleBlockInfo> addPartitionData(int partitionId, byte[] serializedData) {
```
[the remark this answers is not in the document]
Comment (summaryzb) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java:160]: Wrap uniffle writer in gluten side writer. `loop start and loop end` is called in it

File: client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java [hunk 7]
```diff
@@ -151,25 +154,20 @@ public WriteBufferManager(
...
+    return addPartitionData(partitionId, serializedData, 0, 0L);
```
[the remark this answers is not in the document]
Comment (summaryzb) [reply client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java]: No, the record and bytes metric is already counted in gluten side.
